src/games/strategy/triplea/ai/proAI/util/ProOddsCalculator.java

```diff
@@ -0,0 +1,173 @@
+
+    final ProBattleResult result = checkIfNoAttackersOrDefenders(t, attackingUnits, defendingUnits);
+    if (result != null) {
+      return result;
+    }
+
+    // Determine if defenders have no chance
+    final double strengthDifference = ProBattleUtils.estimateStrengthDifference(t, attackingUnits, defendingUnits);
+    if (strengthDifference > 55) {
+      final boolean isLandAndCanOnlyBeAttackedByAir = !t.isWater() && Match.allMatch(attackingUnits, Matches.UnitIsAir);
+      return new ProBattleResult(100 + strengthDifference, 999 + strengthDifference, !isLandAndCanOnlyBeAttackedByAir,
+          attackingUnits, new ArrayList<Unit>(), 1);
+    }
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits);
+  }
+
+  public ProBattleResult calculateBattleResults(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean isAttacker) {
+
+    final ProBattleResult result = checkIfNoAttackersOrDefenders(t, attackingUnits, defendingUnits);
+    if (result != null) {
+      return result;
+    }
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits);
+  }
+
+  private ProBattleResult checkIfNoAttackersOrDefenders(final Territory t, final List<Unit> attackingUnits,
+      final List<Unit> defendingUnits) {
+    final GameData data = ProData.getData();
+
+    final boolean hasNoDefenders = Match.noneMatch(defendingUnits, Matches.UnitIsNotInfrastructure);
+    final boolean isLandAndCanOnlyBeAttackedByAir = !t.isWater() && Match.allMatch(attackingUnits, Matches.UnitIsAir);
+    if (attackingUnits.size() == 0) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders && isLandAndCanOnlyBeAttackedByAir) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders) {
+      return new ProBattleResult(100, 0.1, true, attackingUnits, new ArrayList<Unit>(), 0);
+    } else if (Properties.getSubRetreatBeforeBattle(data) && Match.allMatch(defendingUnits, Matches.UnitIsSub)
+        && Match.noneMatch(attackingUnits, Matches.UnitIsDestroyer)) {
+      return new ProBattleResult();
+    }
+    return null;
+  }
+
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits) {
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits, false);
+  }
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean retreatWhenOnlyAirLeft) {
+    final GameData data = ProData.getData();
+
+    if (isCanceled || attackingUnits.isEmpty() || defendingUnits.isEmpty()) {
```

Comment:
Okay, in this case I think this implementation is a bit broken w.r.t isCancelled. What is the chance that stop game will be called between the time this method is invoked, and the time that this first 'isCanceled' line is executed? My point being, we're *much* more likely to get the 'isCanceled' to be active furtehr down in this method, while we are doing heavy lifting. So the early termination here is a false sense of security.